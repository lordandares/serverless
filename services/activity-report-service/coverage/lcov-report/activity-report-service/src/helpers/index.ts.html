<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for activity-report-service/src/helpers/index.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">activity-report-service/src/helpers</a> index.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>34/34</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>18/18</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20</td><td class="line-coverage quiet"><span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import * as s3 from './s3'
&nbsp;
export { buildLocationSummaryPdf } from './build-location-summary-pdf'
export { buildShiftSummaryPdf } from './build-shift-summary-pdf'
export { buildEmail } from './build-email'
export { buildShiftEmail } from './build-shift-email'
export { getShiftDetails } from './get-shift-details'
export { getShiftReportLinks } from './get-shift-report-links'
export { getUserDetails } from './get-user-details'
export { concatPdfs } from './concat-pdfs'
export { downloadFiles } from './download-files'
export { fetchApplications } from './fetch-applications'
export { fetchLocations } from './fetch-locations'
export { getLocationReportLinks } from './get-location-report-links'
export { getShiftSummaryData } from './get-shift-summary-data'
export { getSummaryData } from './get-summary-data'
export { getFormS3Paths } from './get-references'
export { s3 }
export { sendEmail } from './send-email'
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Jan 13 2020 21:52:44 GMT+0000 (Coordinated Universal Time)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
