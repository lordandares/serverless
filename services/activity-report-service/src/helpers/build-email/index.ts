import * as BPromise from 'bluebird'
import {
  filter,
  find,
  flow,
  get,
  getOr,
  isEmpty,
  reject,
  toUpper,
  trim,
} from 'lodash/fp'
import * as moment from 'moment-timezone'

interface Env {
  env: {
    STAGE: string
    SES_CONFIGURATION_SET_NAME: string
  }
}

declare var process: Env

import { Location } from '../../handlers/email-processor'
import { getLocationReportLinks } from '../get-location-report-links'
import { UserDetails } from '../get-user-details'

const getFormSubmissionOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(true, 'options.formSubmissions'),
)

const getNoActivityOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(false, 'options.skipNoActivity'),
)

export interface BuildEmail extends UserDetails {
  locations: Location[]
}

export async function buildEmail(options: BuildEmail) {
  try {
    const {
      applicationName,
      email,
      firstName,
      locations,
      preferences,
      timezone,
    } = options

    const isProduction = process.env.STAGE === 'production'
    const stage = toUpper(process.env.STAGE)
    const template = isProduction
      ? 'daily-location-reports'
      : 'daily-location-reports-test'

    const ConfigurationSetName = process.env.SES_CONFIGURATION_SET_NAME

    const mStartDate = moment()
      .utc()
      .tz(timezone)
      .subtract(1, 'day')

    const dateText = mStartDate.format('dddd Do MMMM')
    const shortDateText = mStartDate.format('MMMM D')

    const withFormSubmissions = getFormSubmissionOption(preferences)
    const skipNoActivity = getNoActivityOption(preferences)

    const links = await BPromise.map(locations, location =>
      getLocationReportLinks({
        location,
        withFormSubmissions,
      }),
    )

    const activeLocations = filter('link')(links)
    const inactiveLocations = reject('link')(links)

    const hasNoLocations =
      isEmpty(activeLocations) && isEmpty(inactiveLocations)

    if (
      isEmpty(activeLocations) &&
      !isEmpty(inactiveLocations) && skipNoActivity
    ) {
      console.info('skipping activity report', {
        activeLocations: activeLocations.length,
        inactiveLocations: inactiveLocations.length,
        skipNoActivity,
      })
      return false
    }

    const name = trim(firstName)

    const data = JSON.stringify({
      activeLocations,
      applicationName,
      date: dateText,
      hasNoLocations,
      inactiveLocations,
      isProduction,
      name,
      sendInactive: !skipNoActivity,
      shortDate: shortDateText,
      stage,
    })

    return {
      ...(ConfigurationSetName && { ConfigurationSetName }),
      Destination: {
        CcAddresses: [],
        ToAddresses: [email],
      },
      ReplyToAddresses: ['notifications@lighthouse.io'],
      Source: 'notifications@lighthouse.io',
      Template: template,
      TemplateData: data,
    }
  } catch (error) {
    console.error('Error building email', { error, options })
    return false
  }
}
