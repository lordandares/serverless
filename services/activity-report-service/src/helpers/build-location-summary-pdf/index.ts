import { buildActivityPdf } from '@lighthouse/common'
import * as dateFns from 'date-fns'

import { buildConfig, createSummaryPdf } from '../create-summary-pdf'

export async function buildLocationSummaryPdf(data) {
  const config = buildConfig(data)
  const { data: summaryData, pdfOptions } = config

  const pdfDefinition = await buildActivityPdf(pdfOptions, summaryData)
  const s3FullPath = getLocationSummaryPdfPath(data)
  await createSummaryPdf(pdfDefinition, s3FullPath)

  return s3FullPath
}

function getLocationSummaryPdfPath({ location }) {
  const date = new Date()
  const year = dateFns.format(date, 'YYYY')
  const month = dateFns.format(date, 'MM')
  const day = dateFns.format(date, 'DD')
  const unix = date.getTime()

  const { _id: locationId } = location
  const s3PathName = `activity-reports/${locationId}/${year}/${month}`
  const s3FileName = `${year}-${month}-${day}-summary-${unix}.pdf`
  return `${s3PathName}/${s3FileName}`
}
