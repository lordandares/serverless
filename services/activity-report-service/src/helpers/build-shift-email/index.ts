import * as BPromise from 'bluebird'
import {
  filter,
  isEmpty,
  reject,
  sortBy,
  toUpper,
  trim,
  getOr,
  flow,
  find,
  get,
} from 'lodash/fp'
import { orderBy } from 'lodash'
import * as moment from 'moment-timezone'

import { getShiftReportLinks } from '../get-shift-report-links'

interface Env {
  env: {
    STAGE: string
    SES_CONFIGURATION_SET_NAME: string
  }
}

declare var process: Env

const getFormSubmissionOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-shift-report' }),
  getOr(true, 'options.formSubmissions'),
)

export async function buildShiftEmail({ userDetails, datetime }) {
  try {
    const {
      applicationName,
      email,
      firstName,
      preferences,
      timezone,
      shifts,
    } = userDetails

    const isProduction = process.env.STAGE === 'production'
    const stage = toUpper(process.env.STAGE)
    const template = isProduction
      ? 'daily-shift-reports'
      : 'daily-shift-reports-test'

    const ConfigurationSetName = process.env.SES_CONFIGURATION_SET_NAME

    const mStartDate = moment(datetime)
      .utc()
      .tz(timezone)
      .subtract(1, 'day')

    const dateText = mStartDate.format('dddd Do MMMM')
    const shortDateText = mStartDate.format('MMMM D')

    const withFormSubmissions = getFormSubmissionOption(preferences)

    const shiftLinks = await BPromise.map(shifts, shift =>
      getShiftReportLinks({
        datetime,
        shift,
        withFormSubmissions,
      }),
    )

    const activeShifts = orderBy(
      filter('link')(shiftLinks),
      ['lastName', 'timestamp'],
      ['asc'],
    )

    const hasNoShifts = isEmpty(activeShifts)
    const name = trim(firstName)

    const data = JSON.stringify({
      activeShifts,
      applicationName,
      date: dateText,
      hasNoShifts,
      isProduction,
      name,
      shortDate: shortDateText,
      stage,
    })

    return {
      ...(ConfigurationSetName && { ConfigurationSetName }),
      Destination: {
        CcAddresses: [],
        ToAddresses: [email],
      },
      ReplyToAddresses: ['notifications@lighthouse.io'],
      Source: 'notifications@lighthouse.io',
      Template: template,
      TemplateData: data,
    }
  } catch (error) {
    console.error('Error building email', { error, userDetails })
    return false
  }
}
