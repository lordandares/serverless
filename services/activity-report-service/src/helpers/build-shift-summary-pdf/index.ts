import { buildShiftPdf } from '@lighthouse/common'
import * as dateFns from 'date-fns'

import { buildConfig, createSummaryPdf } from '../create-summary-pdf'

export async function buildShiftSummaryPdf(data) {
  const config = buildConfig(data)
  const { data: summaryData, pdfOptions } = config

  const pdfDefinition = await buildShiftPdf(pdfOptions, summaryData)
  const s3FullPath = getShiftSummaryPdfPath(data)
  await createSummaryPdf(pdfDefinition, s3FullPath)

  return s3FullPath
}

function getShiftSummaryPdfPath({ applicationId, shift }) {
  const date = new Date()
  const year = dateFns.format(date, 'YYYY')
  const month = dateFns.format(date, 'MM')
  const day = dateFns.format(date, 'DD')
  const unix = date.getTime()

  const { _id: shiftId } = shift
  const s3PathName = `activity-reports/shifts/${applicationId}/${year}/${month}/${day}`
  const s3FileName = `${year}-${month}-${day}-shift-${shiftId}-summary-${unix}.pdf`
  return `${s3PathName}/${s3FileName}`
}
