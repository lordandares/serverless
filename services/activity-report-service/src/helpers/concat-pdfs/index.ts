import { each } from 'lodash/fp'
import * as uuid from 'node-uuid'

import * as hummus from '@lighthouse/hummus-lambda'

export async function concatPdfs(filePaths, directoryPath) {
  const fileName = 'output.pdf'
  const filePath = `${directoryPath}/${fileName}`

  const pdfWriter = hummus.createWriter(filePath)
  each(path => appendPDFPageFromPDFWithAnnotations(pdfWriter, path), filePaths)
  pdfWriter.end()

  return filePath
}

/* istanbul ignore next */
function appendPDFPageFromPDFWithAnnotations(pdfWriter, sourcePDFPath) {
  const cpyCxt = pdfWriter.createPDFCopyingContext(sourcePDFPath)
  const cpyCxtParser = cpyCxt.getSourceDocumentParser()

  for (let i = 0; i < cpyCxtParser.getPagesCount(); ++i) {
    const pageDictionary = cpyCxtParser.parsePageDictionary(i)

    if (!pageDictionary.exists('Annots')) {
      // no annotation. append as is
      cpyCxt.appendPDFPageFromPDF(i)
    } else {
      // this let here will save any reffed objects from the copied annotations object.
      // they will be written after the page copy writing as to not to disturb the
      // page object writing itself.
      let reffedObjects

      pdfWriter.getEvents().once('OnPageWrite', params => {
        // using the page write event, write the new annotations. just copy the object
        // as is, saving any referenced objects for future writes
        params.pageDictionaryContext.writeKey('Annots')
        reffedObjects = cpyCxt.copyDirectObjectWithDeepCopy(
          pageDictionary.queryObject('Annots'),
        )
      })
      // write page. this will trigger the event
      cpyCxt.appendPDFPageFromPDF(i)
      // now write the reffed object (should be populated cause onPageWrite was written)
      // note that some or all annotations may be embedded, in which case this array
      // wont hold all annotation objects
      if (reffedObjects && reffedObjects.length > 0) {
        cpyCxt.copyNewObjectsForDirectObject(reffedObjects)
      }
    }
  }
}
