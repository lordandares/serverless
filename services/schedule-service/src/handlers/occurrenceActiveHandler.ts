import { errors, schemas } from '@lighthouse/serverless-common'
import { Context, SNSEvent } from 'aws-lambda'
import * as AWS from 'aws-sdk'
import { ClientConfiguration, DocumentClient } from 'aws-sdk/clients/dynamodb'
import * as AWSXRay from 'aws-xray-sdk'
import * as moment from 'moment'

import { OCCURRENCE_STATUS_ACTIVE } from '../constants'

import { payloadToRulePatternDocument } from '../service/lib/ruleTransform'
import parseSnsEvent from './lib/parseSnsEvent'

declare var process: {
  env: {
    IS_OFFLINE: string
    REGION: string
    TABLE_SCHEDULES: string
  }
}

export async function occurrenceActiveHandler(
  event: SNSEvent,
  context: Context,
) {
  try {
    const { body } = parseSnsEvent(event)

    if (!body) {
      throw new errors.ApplicationError({
        message: 'occurrenceActiveHandler: missing body on event',
      })
    }

    const { pk: occurrencePk, sk: occurrenceSk } = body
    const { awsRequestId, invokedFunctionArn } = context

    const isOffline = process.env.IS_OFFLINE
    const region = process.env.REGION
    const tableName = process.env.TABLE_SCHEDULES

    const dynamoDbConfig: ClientConfiguration =
      region === 'localhost'
        ? /* istanbul ignore next */ {
            endpoint: 'http://localhost:8000',
            httpOptions: { timeout: 5000 },
            region,
          }
        : {}

    const DynamoDB = isOffline
      ? AWS.DynamoDB
      : /* istanbul ignore next */
        AWSXRay.captureAWS(AWS).DynamoDB

    const documentClient: DocumentClient = new DynamoDB.DocumentClient(
      dynamoDbConfig,
    )

    const occurrenceGetParams: DocumentClient.GetItemInput = {
      Key: { pk: occurrencePk, sk: occurrenceSk },
      TableName: tableName,
    }

    const occurrenceGetResponse: DocumentClient.GetItemOutput = await documentClient
      .get(occurrenceGetParams)
      .promise()

    const { Item: occurrenceDocument } = occurrenceGetResponse

    if (!occurrenceDocument) {
      console.warn(
        'OccurrenceActiveHandler: The occurrence is missing so unable to add rule or set occurrence to active',
      )
      return
    }

    console.debug('OccurrenceActiveHandler: setting occurrence to active', {
      occurrence: JSON.stringify(occurrenceDocument),
    })

    const {
      applicationId,
      locationId,
      occurrenceId,
      status,
    } = occurrenceDocument

    if (status === OCCURRENCE_STATUS_ACTIVE) {
      console.warn(
        'OccurrenceActiveHandler: occurrence already has active status so skipping',
        {
          occurrenceId,
        },
      )
      return
    }

    const rulePk = `rule-pattern-${applicationId}-${locationId}`
    const ruleSk = 'visit'

    const rulePatternGetParams: DocumentClient.GetItemInput = {
      Key: { pk: rulePk, sk: ruleSk },
      TableName: tableName,
    }

    const rulePatternGetResponse: DocumentClient.GetItemOutput = await documentClient
      .get(rulePatternGetParams)
      .promise()

    const { Item: existingRulePattern } = rulePatternGetResponse

    const actor: schemas.ActorSchema = {
      id: awsRequestId,
      label: invokedFunctionArn,
      type: 'system',
    }

    const rulePayload = {
      applicationId,
      locationId,
      occurrenceId,
      pk: occurrencePk,
      sk: occurrenceSk,
      type: 'visit',
    }

    const patternDocument: schemas.RulePatternDocumentSchema = payloadToRulePatternDocument(
      {
        actor,
        existingDocument: existingRulePattern as schemas.RulePatternDocumentSchema,
        payload: rulePayload,
      },
    )

    try {
      await schemas.validate({
        data: patternDocument,
        schema: schemas.rulePatternDocumentSchema,
      })
    } catch (err) {
      throw new errors.ApplicationError({ message: err.message })
    }

    const transactItems: DocumentClient.TransactWriteItemList = []

    const updatedAtIsoString = moment()
      .utc()
      .toDate()
      .toISOString()

    // update or add new rule pattern document
    if (!existingRulePattern) {
      console.info(
        'OccurrenceActiveHandler: adding new rule pattern document',
        {
          patternDocument,
        },
      )

      transactItems.push({
        Put: {
          Item: patternDocument,
          TableName: tableName,
        },
      })
    } else {
      console.info(
        'OccurrenceActiveHandler: updating existing rule pattern document',
        {
          patternDocument,
        },
      )

      transactItems.push({
        Update: {
          ConditionExpression: '#updatedAt = :expectedUpdatedAt',
          ExpressionAttributeNames: {
            '#matches': 'matches',
            '#updatedAt': 'updatedAt',
            '#updatedBy': 'updatedBy',
          },
          ExpressionAttributeValues: {
            ':expectedUpdatedAt': existingRulePattern.updatedAt,
            ':matches': patternDocument.matches,
            ':nextUpdatedAt': updatedAtIsoString,
            ':updatedBy': actor,
          },
          Key: { pk: patternDocument.pk, sk: patternDocument.sk },
          TableName: tableName,
          UpdateExpression:
            'SET #matches = :matches, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy',
        },
      })
    }

    console.info('OccurrenceActiveHandler: updating existing occurrence', {
      occurrenceDocument,
    })

    // update occurrence so status is now active
    transactItems.push({
      Update: {
        ConditionExpression: '#updatedAt = :expectedUpdatedAt',
        ExpressionAttributeNames: {
          '#status': 'status',
          '#updatedAt': 'updatedAt',
          '#updatedBy': 'updatedBy',
        },
        ExpressionAttributeValues: {
          ':expectedUpdatedAt': occurrenceDocument.updatedAt,
          ':nextUpdatedAt': updatedAtIsoString,
          ':status': OCCURRENCE_STATUS_ACTIVE,
          ':updatedBy': actor,
        },
        Key: { pk: occurrenceDocument.pk, sk: occurrenceDocument.sk },
        TableName: tableName,
        UpdateExpression:
          'SET #status = :status, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy',
      },
    })

    const transactParams: DocumentClient.TransactWriteItemsInput = {
      TransactItems: transactItems,
    }

    const transactItemsLength = transactItems.length

    try {
      console.info(
        `OccurrenceActiveHandler: processing ${transactItemsLength} transactions`,
      )

      const response: DocumentClient.TransactWriteItemsOutput = await documentClient
        .transactWrite(transactParams)
        .promise()

      console.info(
        'OccurrenceActiveHandler: completed transactions successfully',
      )

      return response
    } catch (err) {
      throw new errors.ApplicationError({ message: err.message })
    }
  } catch (err) {
    console.error('OccurrenceActiveError', {
      err,
      event: JSON.stringify(event),
    })

    if (errors.isKnownError(err)) {
      throw err
    }

    throw new errors.UnknownError()
  }
}
