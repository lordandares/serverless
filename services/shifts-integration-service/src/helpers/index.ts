export { default as getPlatformEndpoint } from './get-platform-endpoint'
export { default as getSnsPayload } from './get-sns-payload'
export {
  getWinTeamMessages,
  getWinTeamTimePunch,
} from './get-winteam-time-punch'
export { default as notifyUser } from './notify-user'
export { default as parseWinTeamMessages } from './parse-winteam-messages'
export { default as sendMessage } from './send-message'
export {
  default as validateEndpointAttributes,
} from './validate-endpoint-attributes'
