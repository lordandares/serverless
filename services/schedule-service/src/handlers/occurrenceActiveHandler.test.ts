import { errors, schemas } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk-mock'
import * as MockDate from 'mockdate'

import { createContext, createSnsEvent } from '../../../../__test__/helpers'
import {
  context,
  occurrenceDocument,
  rulePatternDocument,
} from '../__fixtures__'
import { occurrenceActiveHandler } from './occurrenceActiveHandler'

const timestamp = '2019-01-01T01:00:00.000Z'

beforeEach(() => {
  process.env.IS_OFFLINE = 'true'
  process.env.TABLE_SCHEDULES = 'table-schedules'
  MockDate.set(timestamp)
})

afterEach(() => {
  AWS.restore()
  jest.clearAllMocks()
  MockDate.reset()
})

test('application error when body missing', async () => {
  expect.assertions(1)

  const body = null

  try {
    await occurrenceActiveHandler(createSnsEvent({ body }), context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: occurrenceActiveHandler: missing body on event]`,
    )
  }
})

test('unknown error', async () => {
  expect.assertions(1)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }

  const getSpy = jest.fn().mockRejectedValue(new Error('Unknown Error!'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  try {
    await occurrenceActiveHandler(createSnsEvent({ body }), context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[UnknownError: Something went wrong! Try again or contact support if the problem persists.]`,
    )
  }

  AWS.restore()
})

test('returns successfully when a new rule pattern document', async () => {
  expect.assertions(3)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: occurrenceDocument })
    .mockResolvedValueOnce({ Item: undefined })
  const transactSpy = jest.fn().mockResolvedValue({})

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactSpy)

  const result = await occurrenceActiveHandler(
    createSnsEvent({ body }),
    context,
  )

  expect(getSpy.mock.calls).toMatchSnapshot()
  expect(transactSpy.mock.calls).toMatchSnapshot()
  expect(result).toMatchSnapshot()
})

test('returns successfully when updating rule pattern document', async () => {
  expect.assertions(3)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: occurrenceDocument })
    .mockResolvedValueOnce({ Item: rulePatternDocument })
  const transactSpy = jest.fn().mockResolvedValue({})

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactSpy)

  const result = await occurrenceActiveHandler(
    createSnsEvent({ body }),
    context,
  )

  expect(getSpy.mock.calls).toMatchSnapshot()
  expect(transactSpy.mock.calls).toMatchSnapshot()
  expect(result).toMatchSnapshot()
})

test('skips and warns when occurrence missing', async () => {
  expect.assertions(1)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }

  const getSpy = jest.fn().mockResolvedValueOnce({ Item: undefined })
  const warnSpy = jest.spyOn(console, 'warn')

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  await occurrenceActiveHandler(createSnsEvent({ body }), context)

  expect(warnSpy).toHaveBeenCalledWith(
    'OccurrenceActiveHandler: The occurrence is missing so unable to add rule or set occurrence to active',
  )
})

test('skips and warns when occurrence is already active', async () => {
  expect.assertions(1)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({
      Item: { ...occurrenceDocument, status: 'active' },
    })
  const warnSpy = jest.spyOn(console, 'warn')

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  await occurrenceActiveHandler(createSnsEvent({ body }), context)

  expect(warnSpy).toHaveBeenCalledWith(
    'OccurrenceActiveHandler: occurrence already has active status so skipping',
    {
      occurrenceId: 'occurrence1',
    },
  )
})

test('application error when transactions fail', async () => {
  expect.assertions(1)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: occurrenceDocument })
    .mockResolvedValueOnce({ Item: rulePatternDocument })
  const transactSpy = jest
    .fn()
    .mockRejectedValue(new Error('Transaction Error!'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactSpy)

  try {
    await occurrenceActiveHandler(createSnsEvent({ body }), context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: Transaction Error!]`,
    )
  }

  AWS.restore()
})

test('patternDocument validation error', async () => {
  expect.assertions(1)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: occurrenceDocument })
    .mockResolvedValueOnce({ Item: null })

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  schemas.validate = jest.fn().mockRejectedValue(new Error('Validation Error'))

  try {
    await occurrenceActiveHandler(createSnsEvent({ body }), context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(`[ApplicationError: Validation Error]`)
  }
})
