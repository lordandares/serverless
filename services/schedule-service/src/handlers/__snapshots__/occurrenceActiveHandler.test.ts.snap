// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns successfully when a new rule pattern document 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "pk": "application1-occurrence",
        "sk": "2019-01-01T00:00-occurrence1",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
  Array [
    Object {
      "Key": Object {
        "pk": "rule-pattern-application1-5d163137b8b3b7000127edd1",
        "sk": "visit",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`returns successfully when a new rule pattern document 2`] = `
Array [
  Array [
    Object {
      "TransactItems": Array [
        Object {
          "Put": Object {
            "Item": Object {
              "createdAt": "2019-01-01T01:00:00.000Z",
              "createdBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
              "groupType": "rule-pattern",
              "matches": Object {
                "occurrence1": Object {
                  "pk": "application1-occurrence",
                  "sk": "2019-01-01T00:00-occurrence1",
                },
              },
              "pk": "rule-pattern-application1-5d163137b8b3b7000127edd1",
              "sk": "visit",
              "updatedAt": "2019-01-01T01:00:00.000Z",
              "updatedBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
            },
            "TableName": "table-schedules",
          },
        },
        Object {
          "Update": Object {
            "ConditionExpression": "#updatedAt = :expectedUpdatedAt",
            "ExpressionAttributeNames": Object {
              "#status": "status",
              "#updatedAt": "updatedAt",
              "#updatedBy": "updatedBy",
            },
            "ExpressionAttributeValues": Object {
              ":expectedUpdatedAt": undefined,
              ":nextUpdatedAt": "2019-01-01T01:00:00.000Z",
              ":status": "active",
              ":updatedBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
            },
            "Key": Object {
              "pk": "application1-occurrence",
              "sk": "2019-01-01T00:00-occurrence1",
            },
            "TableName": "table-schedules",
            "UpdateExpression": "SET #status = :status, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy",
          },
        },
      ],
    },
    [Function],
  ],
]
`;

exports[`returns successfully when a new rule pattern document 3`] = `Object {}`;

exports[`returns successfully when updating rule pattern document 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "pk": "application1-occurrence",
        "sk": "2019-01-01T00:00-occurrence1",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
  Array [
    Object {
      "Key": Object {
        "pk": "rule-pattern-application1-5d163137b8b3b7000127edd1",
        "sk": "visit",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`returns successfully when updating rule pattern document 2`] = `
Array [
  Array [
    Object {
      "TransactItems": Array [
        Object {
          "Update": Object {
            "ConditionExpression": "#updatedAt = :expectedUpdatedAt",
            "ExpressionAttributeNames": Object {
              "#matches": "matches",
              "#updatedAt": "updatedAt",
              "#updatedBy": "updatedBy",
            },
            "ExpressionAttributeValues": Object {
              ":expectedUpdatedAt": "2019-06-02T00:00:00.000Z",
              ":matches": Object {
                "occurrence1": Object {
                  "pk": "application1-occurrence",
                  "sk": "2019-01-01T00:00-occurrence1",
                },
              },
              ":nextUpdatedAt": "2019-01-01T01:00:00.000Z",
              ":updatedBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
            },
            "Key": Object {
              "pk": "rule-pattern-application1-5d163137b8b3b7000127edd1",
              "sk": "visit",
            },
            "TableName": "table-schedules",
            "UpdateExpression": "SET #matches = :matches, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy",
          },
        },
        Object {
          "Update": Object {
            "ConditionExpression": "#updatedAt = :expectedUpdatedAt",
            "ExpressionAttributeNames": Object {
              "#status": "status",
              "#updatedAt": "updatedAt",
              "#updatedBy": "updatedBy",
            },
            "ExpressionAttributeValues": Object {
              ":expectedUpdatedAt": undefined,
              ":nextUpdatedAt": "2019-01-01T01:00:00.000Z",
              ":status": "active",
              ":updatedBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
            },
            "Key": Object {
              "pk": "application1-occurrence",
              "sk": "2019-01-01T00:00-occurrence1",
            },
            "TableName": "table-schedules",
            "UpdateExpression": "SET #status = :status, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy",
          },
        },
      ],
    },
    [Function],
  ],
]
`;

exports[`returns successfully when updating rule pattern document 3`] = `Object {}`;
